ldap_sync/models/res_company_ldap.py

```diff
@@ -0,0 +1,175 @@
+        comodel_name='res.company.ldap.group',
+        inverse_name='company_ldap_id',
+    )
+    group_mapping_ids = fields.One2many(
+        string='LDAP Group Mappings',
+        comodel_name='res.company.ldap.group.mapping',
+        inverse_name='company_ldap_id',
+    )
+
+    @api.multi
+    def _scheduled_sync(self):
+        try:
+            self.filtered('scheduled_sync_enabled').with_context(
+                scheduled=True,
+            )._sync()
+        except ldap.LDAPError as e:
+            _logger.error('LDAP exception occurred: %s', e)
+
+    @api.multi
+    def _sync(self):
+        for server in self:
+            server._sync_groups()
+            server._sync_users()
+
+    @api.multi
+    def _get_sync_query_config(self, base_dn):
+        self.ensure_one()
+        config = self._get_ldap_dicts()[0]
+        config.update({
+            'ldap_base': base_dn,
+        })
+        return config
+
+    @api.multi
+    def _get_sync_group_attribs(self):
+        self.ensure_one()
+        attribs = [
+            self.group_dn_attribute,
+            self.group_name_attribute,
+        ]
+
+    @api.multi
+    def _prepare_group_values(self, attribs):
+        self.ensure_one()
+        return {
+            'dn': attribs[self.group_dn_attribute][0].decode(),
+            'name': attribs[self.group_name_attribute][0].decode(),
+        }
+
+    @api.multi
+    def _sync_groups(self):
+        self.ensure_one()
+        ResCompanyLdapGroup = self.env['res.company.ldap.group']
+        results = self._query(
+            self._get_sync_query_config(self.groups_dn),
+            tools.ustr(self.group_object_filter),
+            retrieve_attributes=self._get_sync_group_attribs(),
+        )
+        group_ids = [(5, False, False)]
+        for entry in results:
+            if not entry[0]:
+                continue
+            attribs = entry[1]
+            group_id = ResCompanyLdapGroup.search(
+                [
+                    ('company_ldap_id', '=', self.id),
+                    ('dn', '=', attribs[self.group_dn_attribute][0].decode()),
+                ],
+                limit=1,
+            )
+            if group_id:
+                group_ids.append(
+                    (4, group_id.id, False)
+                )
+            else:
+                group_ids.append(
+                    (0, False, self._prepare_group_values())
+                )
+        self.group_ids = group_ids
+
+    @api.multi
+    def _sync_users(self):
+        self.ensure_one()
+        # 1. query all users from LDAP
+        # self._query({ ldap_base: users_dn }, user_object_filter)
+
+        # 2. for all users that are new in LDAP results, look same in existing database
+        #     - link them
+
+        # 3. for each of existing user, by ldap unique ID, select from ldap results and:
+        #     - update if exists: attributes, group membership
+        #     - deactivate if no longer exists and remove link to LDAP
```

Comment:
is something missing here?

Reply:
Yes, still WIP, should be done in an hour or so

Reply:
Well, I underestimated amount of work, but fine now :)